SLING-12714 - Impementation of OidcAuthenticationHandler

nscendoni opened the pull request:
First Implementation of:

- OidcAuthenticationHandler: Sling Authentication Handler
- OidcIdentityProvider: Oak External Identity Provider to relay on Oak DeafaultSyncHandler for user provisioning, including group membership
- Sling Cookie set/verified for pure Sling Setup (analog to Sling org-apache-sling-auth-form and org-apache-sling-auth-saml2)
- Exported several interfaces as SPI to allow different implementations
- IT Tests for OIDC Authorization Code flow login in Sling

rombert commented:
Hi @nscendoni  - can the PKCE support be done as a separate PR/Jira issue?

rombert commented on pom.xml:
```diff
@@ -48,7 +48,15 @@
 	<build>
 		<plugins>
-			<plugin>
+            <plugin>
```
What is this supposed to achieve? If you want to export the package just give it a `@Version` annotation, the rest is handled by the configuration inherited from the parent pom.

rombert commented on pom.xml:
```diff
@@ -221,12 +229,16 @@
                                 <repositoryUrl>file://${project.build.directory}/artifacts</repositoryUrl>
                             </repositoryUrls>
                             <launcherArguments>
+                                <vm-options>
```
Please make this part of a profile if needed otherwise we may get into port clashes, especially on Jenkins.

nscendoni replied:
I commented the vm-options. I use them when I start sling with `mvn feature-launcher:start feature-launcher:stop -Dfeature-launcher.waitForInput=true`. It looks to me that a new profile cannot help in this situation when we do not need to make a real build. When added a new `debug` profile and started Sling with: `mvn feature-launcher:start feature-launcher:stop -Dfeature-launcher.waitForInput=true -Pdebug` the additional parameters were not taken.

rombert commented on pom.xml:
```diff
@@ -221,12 +229,16 @@
                             </launcherArguments>
                             <environmentVariables>
                                 <IT_ENCRYPTION_PASSWORD>random</IT_ENCRYPTION_PASSWORD>
+                                <JAVA_OPS>-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=*:3007</JAVA_OPS>
```
Please make this part of a profile if needed otherwise we may get into port clashes, especially on Jenkins.

rombert commented on pom.xml:
```diff
@@ -397,5 +409,47 @@
+            <scope>provided</scope>
+            <version>1.10.0</version>
+        </dependency>
+        <dependency>
```
We should not use the old Felix SCR annotations.

rombert commented on src/main/features/main.json:
```diff
@@ -16,6 +16,9 @@
         "start-order": 24
     },
+    {
+      "id": "com.nimbusds:nimbus-jose-jwt:10.0.2"
```
Please add a start-order, for consistency.

rombert reviewed (changes requested):
I started reviewing and had some initial comments but then saw the move of impl to spi and I stopped :-) 

I am usually very careful about exposing APIs, can you please explain why all of them need to be public and reusable? We may need to create additional abstractions to make sure they make sense to consumers.

nscendoni commented:
> I started reviewing and had some initial comments but then saw the move of impl to spi and I stopped :-)
> 
> I am usually very careful about exposing APIs, can you please explain why all of them need to be public and reusable? We may need to create additional abstractions to make sure they make sense to consumers.

I agree. I plan to rollback the move of impl to spi as soon as I can split spi and implementation for sling-commons-crypto.

nscendoni commented:
> Hi @nscendoni - can the PKCE support be done as a separate PR/Jira issue?

I reverted that change and moved it to a separate branch in our repository. I'll make a new PR once this is consolidated.

rombert commented:
> I agree. I plan to rollback the move of impl to spi as soon as I can split spi and implementation for [sling-commons-crypto](https://issues.apache.org/jira/browse/SLING-commons-crypto).

Ok, then I'll consider this a WIP and not perform a review :-)

anchela commented on src/main/java/org/apache/sling/auth/oauth_client/impl/JcrUserHomeOAuthTokenStore.java:
```diff
@@ -43,15 +43,15 @@
 
     private static final String PROPERTY_NAME_EXPIRES_AT = "expires_at";
-    private static final String PROPERTY_NAME_ACCESS_TOKEN = "access_token";
+    protected static final String PROPERTY_NAME_ACCESS_TOKEN = "access_token";
```
hi @nscendoni , what is the reason for making this constant 'protected'?

anchela commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -0,0 +1,509 @@
+    // We don't want leave the cookie lying around for a long time because it it not needed.
+    // At the same time, some OAuth user authentication flows take a long time due to
+    // consent, account selection, 2FA, etc so we cannot make this too short.
+    protected static final int COOKIE_MAX_AGE_SECONDS = 300;
```
ideally the constants should be private and if the default constants needs to be readable/editable -> introduce a public configuration option for the service.

anchela replied:
i think i made it private... hope that's ok

anchela commented on src/main/java/org/apache/sling/auth/oauth_client/impl/OidcAuthenticationHandler.java:
```diff
@@ -0,0 +1,509 @@
+
+
+    @Override
+    public AuthenticationInfo extractCredentials(@Nullable HttpServletRequest request, @Nullable HttpServletResponse response) {
```
i would suggest to reduce the complexity of this method in order to make it easier to read and maintain.

anchela replied:
i gave it a try.... @nscendoni , please check if i didn't introduce any bugs :)

anchela reviewed (changes requested):
hi @nscendoni as a very first observation before digging deep into the review: i would consistently add notnull/nullable annotations to all method params and return values. IMHO this helps overall code quality and maintainability as the contract on what information is passed and is obtained on return much easier to understand.

anchela reviewed (approved):
my findings are addressed
but there is still the question about the classes in the SPI package from @rombert . how should be proceed to unblock you?

anchela commented:
@rombert , in order to unblock @nscendoni , would you be able to have a quick look if your concerns regarding the spi package have been addressed?
it now is limited to key extension points.
if you don't have capacity to review it, would you be ok with me merging the changes to reduce the number of active branches. we don't plan to release the bundle at this point as we are still in active development and can still come back to the spi question again.
wdyt?

rombert commented:
> @rombert , in order to unblock @nscendoni , would you be able to have a quick look if your concerns regarding the spi package have been addressed? it now is limited to key extension points. if you don't have capacity to review it, would you be ok with me merging the changes to reduce the number of active branches. we don't plan to release the bundle at this point as we are still in active development and can still come back to the spi question again. wdyt?

Yes, no sense in waiting for me now as I can't set aside the time for a larger review.

rombert reviewed (commented):
Fine for me to merge now.
